security/unittest/csrtest.cpp

```diff
@@ -165,6 +165,145 @@ TEST_F(TestCSRWithDevice, Validate256)
   mbedtls_x509_csr_free(&csr);
 }
 
+class TestCSRWithDeviceTPM : public TestCSRWithDevice {
+public:
+  void SetUp() override
+  {
+    static oc_pki_pk_functions_t pk_functions;
+    pk_functions.mbedtls_pk_ecp_gen_key =
+      TestCSRWithDeviceTPM::simulate_tpm_mbedtls_pk_ecp_gen_key;
+    pk_functions.mbedtls_pk_parse_key =
+      TestCSRWithDeviceTPM::simulate_tpm_mbedtls_pk_parse_key;
+    pk_functions.mbedtls_pk_write_key_der =
+      TestCSRWithDeviceTPM::simulate_tpm_mbedtls_pk_write_key_der;
+    memset(&public_key[0], 0, public_key.size());
+    public_key_size = 0;
+    oc_pki_set_pk_functions(&pk_functions);
+    TestCSRWithDevice::SetUp();
+  }
+
+  void TearDown() override
+  {
+    TestCSRWithDevice::TearDown();
+    oc_pki_set_pk_functions(nullptr);
+    public_key_size = 0;
+    memset(&public_key[0], 0, public_key.size());
+  }
+
+protected:
+  static std::array<uint8_t, OC_ECDSA_PUBKEY_SIZE> public_key;
+  static size_t public_key_size;
+  static std::array<uint8_t, OC_ECDSA_PRIVKEY_SIZE> private_key;
+  static size_t private_key_size;
+
+  static int simulate_tpm_mbedtls_pk_parse_key(
+    size_t device, mbedtls_pk_context *pk, const unsigned char *key,
+    size_t keylen, const unsigned char *pwd, size_t pwdlen,
+    int (*f_rng)(void *, unsigned char *, size_t), void *p_rng)
+  {
+    (void)device;
```

Comment:
the `device` parameter is never used for `mbedtls_pk_parse_key_cb_t` nor for `mbedtls_pk_ecp_gen_key_cb_t`, why do we think it's necessary? wouldn't perhaps a `void*` argument for custom user data be more useful?

Reply:
`void*` is not necessary, because pk functions are global, the `device` is used to determine which device asks for the certificate, in case you have multiple devices.